File: clients/macos/Sources/ChatOSConnector/SQLiteProjectRegistry.swift
import ChatOSCore
import Foundation
import SQLite3

/// Account-scoped, local-only authority. Opening a corrupt database fails closed; it never
/// becomes an empty registry. No API client, connector pairing or plugin installation is needed.
public actor SQLiteProjectRegistry: ProjectRegistry {
    private nonisolated(unsafe) var database: OpaquePointer?

    public init(databaseURL: URL) throws {
        try FileManager.default.createDirectory(
            at: databaseURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        var handle: OpaquePointer?
        guard sqlite3_open_v2(databaseURL.path, &handle, SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX, nil) == SQLITE_OK else {
            let message = handle.map { String(cString: sqlite3_errmsg($0)) } ?? "open failed"
            sqlite3_close(handle)
            throw ProjectRegistryError.storage(message)
        }
        do {
            sqlite3_busy_timeout(handle, 5_000)
            guard sqlite3_exec(handle, Self.schema, nil, nil, nil) == SQLITE_OK else {
                throw ProjectRegistryError.storage(String(cString: sqlite3_errmsg(handle)))
            }
            database = handle
        } catch {
            sqlite3_close(handle)
            throw error
        }
    }

    deinit { sqlite3_close(database) }

    public func list(ownerUserID: String, includeInactive: Bool = false) throws -> [LocalProjectRecord] {
        try ProjectRegistryValidation.identifier(ownerUserID, field: "ownerUserID")
        return try query(
            "SELECT \(Self.columns) FROM local_project_records WHERE owner_user_id = ?"
                + (includeInactive ? "" : " AND status = 'active'") + " ORDER BY name, id",
            [.text(ownerUserID)], row: readRecord
        )
    }

    public func get(ownerUserID: String, id: String) throws -> LocalProjectRecord? {
        try ProjectRegistryValidation.identifier(ownerUserID, field: "ownerUserID")
        try ProjectRegistryValidation.identifier(id, field: "id")
        return try query(
            "SELECT \(Self.columns) FROM local_project_records WHERE owner_user_id = ? AND id = ?",
            [.text(ownerUserID), .text(id)], row: readRecord
        ).first
    }

    public func create(ownerUserID: String, draft: LocalProjectDraft) throws -> LocalProjectRecord {
        let now = Self.now()
        let record = LocalProjectRecord(
            id: UUID().uuidString.lowercased(), ownerUserID: ownerUserID, draft: draft,
            createdAtUnixMs: now, updatedAtUnixMs: now
        )
        try record.validate()
        try insert(record)
        return record
    }

    public func update(
        ownerUserID: String, id: String, expectedRevision: Int64,
        draft: LocalProjectDraft, status: LocalProjectStatus
    ) throws -> LocalProjectRecord {
        try transaction {
            guard let old = try get(ownerUserID: ownerUserID, id: id) else { throw ProjectRegistryError.notFound }
            guard old.revision == expectedRevision else { throw ProjectRegistryError.revisionConflict }
            guard old.status != .removed else { throw ProjectRegistryError.removed }
            let record = LocalProjectRecord(
                id: id, ownerUserID: ownerUserID, draft: draft, revision: old.revision + 1,
                status: status, createdAtUnixMs: old.createdAtUnixMs,
                updatedAtUnixMs: max(Self.now(), old.updatedAtUnixMs)
            )
            try record.validate()
            try execute("""
                UPDATE local_project_records SET name = ?, description = ?, workspace_id = ?,
                relative_root = ?, revision = ?, status = ?, updated_at_unix_ms = ?
                WHERE owner_user_id = ? AND id = ? AND revision = ?
                """, [
                    .text(draft.name), .text(draft.description), .text(draft.workspaceID), .text(draft.relativeRoot),
                    .integer(record.revision), .text(status.rawValue), .integer(record.updatedAtUnixMs),
                    .text(ownerUserID), .text(id), .integer(expectedRevision),
                ])
            guard sqlite3_changes(database) == 1 else { throw ProjectRegistryError.revisionConflict }
            return record
        }
    }

    private func insert(_ record: LocalProjectRecord) throws {
        try execute("INSERT INTO local_project_records (\(Self.columns)) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)", [
            .text(record.ownerUserID), .text(record.id), .text(record.draft.name), .text(record.draft.description),
            .text(record.draft.workspaceID), .text(record.draft.relativeRoot), .integer(record.revision),
            .text(record.status.rawValue), .integer(record.createdAtUnixMs), .integer(record.updatedAtUnixMs),
        ])
    }

    private func readRecord(_ statement: OpaquePointer) throws -> LocalProjectRecord {
        guard let status = LocalProjectStatus(rawValue: Self.string(statement, 7)) else {
            throw ProjectRegistryError.storage("invalid project status")
        }
        let record = LocalProjectRecord(
            id: Self.string(statement, 1), ownerUserID: Self.string(statement, 0),
            draft: LocalProjectDraft(name: Self.string(statement, 2), description: Self.string(statement, 3),
                                     workspaceID: Self.string(statement, 4), relativeRoot: Self.string(statement, 5)),
            revision: sqlite3_column_int64(statement, 6), status: status,
            createdAtUnixMs: sqlite3_column_int64(statement, 8), updatedAtUnixMs: sqlite3_column_int64(statement, 9)
        )
        try record.validate()
        return record
    }

    private enum Value { case text(String), integer(Int64) }

    private func query<T>(_ sql: String, _ values: [Value] = [], row: (OpaquePointer) throws -> T) throws -> [T] {
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(database, sql, -1, &statement, nil) == SQLITE_OK, let statement else { throw storageError() }
        defer { sqlite3_finalize(statement) }
        for (offset, value) in values.enumerated() {
            let index = Int32(offset + 1)
            let result: Int32
            switch value {
            case let .text(text):
                result = sqlite3_bind_text(statement, index, text, -1, unsafeBitCast(-1, to: sqlite3_destructor_type.self))
            case let .integer(number): result = sqlite3_bind_int64(statement, index, number)
            }
            guard result == SQLITE_OK else { throw storageError() }
        }
        var rows: [T] = []
        while true {
            switch sqlite3_step(statement) {
            case SQLITE_ROW: rows.append(try row(statement))
            case SQLITE_DONE: return rows
            default: throw storageError()
            }
        }
    }

    private func execute(_ sql: String, _ values: [Value] = []) throws {
        let _: [Int] = try query(sql, values) { _ in 0 }
    }

    private func transaction<T>(_ body: () throws -> T) throws -> T {
        try execute("BEGIN IMMEDIATE")
        do {
            let result = try body()
            try execute("COMMIT")
            return result
        } catch {
            try? execute("ROLLBACK")
            throw error
        }
    }

    private func storageError() -> ProjectRegistryError {
        .storage(String(cString: sqlite3_errmsg(database)))
    }

    private static func string(_ statement: OpaquePointer, _ index: Int32) -> String {
        String(cString: sqlite3_column_text(statement, index))
    }

    private static func now() -> Int64 { Int64(Date().timeIntervalSince1970 * 1_000) }

    private static let columns = "owner_user_id, id, name, description, workspace_id, relative_root, revision, status, created_at_unix_ms, updated_at_unix_ms"
    private static let schema = """
        PRAGMA journal_mode = WAL;
        BEGIN IMMEDIATE;
        CREATE TABLE IF NOT EXISTS local_project_records (
            owner_user_id TEXT NOT NULL, id TEXT NOT NULL, name TEXT NOT NULL,
            description TEXT NOT NULL, workspace_id TEXT NOT NULL, relative_root TEXT NOT NULL,
            revision INTEGER NOT NULL CHECK(revision > 0),
            status TEXT NOT NULL CHECK(status IN ('active', 'archived', 'removed')),
            created_at_unix_ms INTEGER NOT NULL, updated_at_unix_ms INTEGER NOT NULL,
            PRIMARY KEY(owner_user_id, id)
        );
        CREATE TABLE IF NOT EXISTS local_project_schema_migrations (version INTEGER PRIMARY KEY NOT NULL);
        INSERT OR IGNORE INTO local_project_schema_migrations(version) VALUES (1);
        COMMIT;
        """
}

File: clients/windows/src/ChatOS.Connector/Persistence/SqliteProjectRegistry.cs
using ChatOS.Core.Abstractions;
using ChatOS.Core.Domain;
using Microsoft.Data.Sqlite;

namespace ChatOS.Connector.Persistence;

/// <summary>Local-only, account-scoped authority. Database failures propagate, never become an empty list.</summary>
public sealed class SqliteProjectRegistry(LocalStateDatabase database) : IProjectRegistry
{
    private const string Columns = "owner_user_id, id, name, description, workspace_id, relative_root, revision, status, created_at_unix_ms, updated_at_unix_ms";

    public async Task<IReadOnlyList<LocalProjectRecord>> ListAsync(
        string ownerUserId, bool includeInactive = false, CancellationToken cancellationToken = default)
    {
        ProjectRegistryValidation.Identifier(ownerUserId, nameof(ownerUserId));
        await using var connection = await database.OpenConnectionAsync(cancellationToken).ConfigureAwait(false);
        using var command = Command(connection, null,
            $"SELECT {Columns} FROM local_project_records WHERE owner_user_id = @p0" +
            (includeInactive ? "" : " AND status = 'active'") + " ORDER BY name, id", ownerUserId);
        await using var reader = await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
        var records = new List<LocalProjectRecord>();
        while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false)) records.Add(ReadRecord(reader));
        return records;
    }

    public async Task<LocalProjectRecord?> GetAsync(
        string ownerUserId, string id, CancellationToken cancellationToken = default)
    {
        await using var connection = await database.OpenConnectionAsync(cancellationToken).ConfigureAwait(false);
        return await ReadAsync(connection, null, ownerUserId, id, cancellationToken).ConfigureAwait(false);
    }

    public async Task<LocalProjectRecord> CreateAsync(
        string ownerUserId, LocalProjectDraft draft, CancellationToken cancellationToken = default)
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var record = new LocalProjectRecord(Guid.NewGuid().ToString("D"), ownerUserId, draft, 1,
            LocalProjectStatus.Active, now, now);
        record.Validate();
        await using var connection = await database.OpenConnectionAsync(cancellationToken).ConfigureAwait(false);
        await InsertAsync(connection, null, record, cancellationToken).ConfigureAwait(false);
        return record;
    }

    public async Task<LocalProjectRecord> UpdateAsync(
        string ownerUserId, string id, long expectedRevision, LocalProjectDraft draft,
        LocalProjectStatus status, CancellationToken cancellationToken = default)
    {
        await using var connection = await database.OpenConnectionAsync(cancellationToken).ConfigureAwait(false);
        using var transaction = connection.BeginTransaction(deferred: false);
        var old = await ReadAsync(connection, transaction, ownerUserId, id, cancellationToken).ConfigureAwait(false)
            ?? throw new ProjectRegistryException(ProjectRegistryError.NotFound, "Local project not found.");
        if (old.Revision != expectedRevision)
            throw new ProjectRegistryException(ProjectRegistryError.RevisionConflict, "Project has changed. Refresh and retry.");
        if (old.Status == LocalProjectStatus.Removed)
            throw new ProjectRegistryException(ProjectRegistryError.Removed, "Removed projects cannot be updated or restored.");
        var record = old with
        {
            Draft = draft, Status = status, Revision = old.Revision + 1,
            UpdatedAtUnixMs = Math.Max(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(), old.UpdatedAtUnixMs),
        };
        record.Validate();
        using var command = Command(connection, transaction, """
            UPDATE local_project_records SET name = @p0, description = @p1, workspace_id = @p2,
                relative_root = @p3, revision = @p4, status = @p5, updated_at_unix_ms = @p6
            WHERE owner_user_id = @p7 AND id = @p8 AND revision = @p9
            """, draft.Name, draft.Description, draft.WorkspaceId, draft.RelativeRoot,
            record.Revision, StatusText(status), record.UpdatedAtUnixMs, ownerUserId, id, expectedRevision);
        if (await command.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false) != 1)
            throw new ProjectRegistryException(ProjectRegistryError.RevisionConflict, "Project has changed.");
        await transaction.CommitAsync(cancellationToken).ConfigureAwait(false);
        return record;
    }

    private static async Task<LocalProjectRecord?> ReadAsync(
        SqliteConnection connection, SqliteTransaction? transaction,
        string ownerUserId, string id, CancellationToken cancellationToken)
    {
        ProjectRegistryValidation.Identifier(ownerUserId, nameof(ownerUserId));
        ProjectRegistryValidation.Identifier(id, nameof(id));
        using var command = Command(connection, transaction,
            $"SELECT {Columns} FROM local_project_records WHERE owner_user_id = @p0 AND id = @p1", ownerUserId, id);
        await using var reader = await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
        return await reader.ReadAsync(cancellationToken).ConfigureAwait(false) ? ReadRecord(reader) : null;
    }

    private static async Task InsertAsync(
        SqliteConnection connection, SqliteTransaction? transaction, LocalProjectRecord record, CancellationToken cancellationToken)
    {
        using var command = Command(connection, transaction,
            $"INSERT INTO local_project_records ({Columns}) VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7, @p8, @p9)",
            record.OwnerUserId, record.Id, record.Draft.Name, record.Draft.Description,
            record.Draft.WorkspaceId, record.Draft.RelativeRoot, record.Revision, StatusText(record.Status),
            record.CreatedAtUnixMs, record.UpdatedAtUnixMs);
        await command.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    private static LocalProjectRecord ReadRecord(SqliteDataReader reader)
    {
        var status = reader.GetString(7) switch
        {
            "active" => LocalProjectStatus.Active,
            "archived" => LocalProjectStatus.Archived,
            "removed" => LocalProjectStatus.Removed,
            _ => throw new InvalidDataException("Invalid project status."),
        };
        var record = new LocalProjectRecord(reader.GetString(1), reader.GetString(0),
            new(reader.GetString(2), reader.GetString(4), reader.GetString(5), reader.GetString(3)),
            reader.GetInt64(6), status, reader.GetInt64(8), reader.GetInt64(9));
        record.Validate();
        return record;
    }

    private static string StatusText(LocalProjectStatus status) => status.ToString().ToLowerInvariant();

    private static SqliteCommand Command(
        SqliteConnection connection, SqliteTransaction? transaction, string sql, params object[] values)
    {
        var command = connection.CreateCommand();
        command.Transaction = transaction;
        command.CommandText = sql;
        for (var index = 0; index < values.Length; index++) command.Parameters.AddWithValue($"@p{index}", values[index]);
        return command;
    }
}
